### DISCUSSION TITLE
Why user_proxy executes code, but not assistant in fomous demo example?

### DISCUSSION THREAD

**Discussion by alexey2baranov** [microsoft/autogen#1730] [General]
From the human point of view it's strange if Human ask " Which big tech stock has the largest year-to-date gain this year?", the AI assistant returns code and ask Human run it. Why AI assistant can't run the code himself ?? Does user_proxy ask "Give me a CODE which let me know answer" ?? NO. He asks "Which company" it means give me a string which contains name of company.

This violates separation of concern ideology imho. Assistant supposed to have all necessary tools to complete task. And user_proxy doesn't care how AI assistant get the answer. Why user_proxy must have python executor instead of this? In such simple hierarchy it's quite easy to manage. But in real complex agent graph such cross dependency will limit ease of connection new agents as we are not able just to connect special agent to chat as we do in real human world with human assistant, but we also should modify the other peace of our code.

The correct way of such conversation imho 
1.  user_proxy asks AI Assistant stock questions
2. AI Assistant perform all neccessary operation for complete task (write code and executes code iteratively)
3. AI Assistant return STRING answer to user_proxy.

What do you think guys. I'm interesting your mention.

**Comment by alexey2baranov** [comment]
BTW a couple words about repeation in this scenario.
 
AI Assistant can be guided in system message or task message to run code iteratively not more then 10 times in order to control token usage. And if he iterates 10 times and doesn't get the final result, he responds "Hey mister , I'm sorry I can't provide an answer with your limit of 10 attempts". So in that moment user_proxy asks a real User the input. And if User want, he writes "Let's proceed with this task 5 times more".

That's the correct way of handling iterations and token usage from human point of view.

**Reply by sonichi** [reply]
This can be achieved by using a "society-of-mind" agent which contains nested chat inside an agent. There is an example at https://microsoft.github.io/autogen/docs/notebooks/agentchat_society_of_mind. @qingyun-wu is working on a PR to make it easier to register nested chat for any agent.

**Comment by alexey2baranov** [comment]
@sonichi  thanks for pointing. i will check this .

Could you tell is is possible technically to do coding/debugging with a single assistant agent? The reason why I'm asking is that from my point of view coding and debugging is a single responsibility of Coder assistant, so I would like to keep this work inside one agent. In agent swarm such agent may write/debug code and provide the result of his work (code) to other agents. 

So is it possible to write such single agent in Autogen? As I understand I can write a system prompt like "provide a Python code. before answer try to execute it to ensure it works, using your pyton enterpretor tool. If it fails fix it. Repeate this steps 10 times, or as much as the user order you." 

Tell me please  is it technically possible ?
